File: 2020/day01.py
import itertools
import math


def part1():
    for expenses in itertools.combinations(read_expenses(), 2):
        if sum(expenses) == 2020:
            return math.prod(expenses)


def part2():
    for expenses in itertools.combinations(read_expenses(), 3):
        if sum(expenses) == 2020:
            return math.prod(expenses)


def read_expenses():
    with open('day01.txt') as file:
        for line in file:
            yield int(line.strip())


def main():
    print(part1())
    print(part2())


if __name__ == '__main__':
    main()

File: go/2024/day01.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	fi, err := os.Open("day01.txt")

	if err != nil {
		panic(err)
	}

	defer fi.Close()

	scanner := bufio.NewScanner(fi)
	var left []int
	var right []int

	for scanner.Scan() {
		line := scanner.Text()
		pair := strings.Split(line, " ")
		p1, _ := strconv.Atoi(pair[0])
		p2, _ := strconv.Atoi(pair[3])
		left = append(left, p1)
		right = append(right, p2)
	}

	sort.Slice(left, func(i, j int) bool { return left[i] < left[j] })
	sort.Slice(right, func(i, j int) bool { return right[i] < right[j] })

	var diffs []float64
	for i, val := range left {
		diffs = append(diffs, math.Abs(float64(val-right[i])))
	}

	var sum float64 = 0
	for _, val := range diffs {
		sum += val
	}

	fmt.Println(int(sum))

	if err := scanner.Err(); err != nil {
		panic(err)
	}
}
